src/aiida_quantumespresso/workflows/xspectra/crystal.py

```diff
@@ -0,0 +1,676 @@
+            num_core_states = int(header_line.split(' ')[5])
+            if num_core_states == 0:
+                return self.exit_codes.ERROR_NO_GIPAW_INFO_FOUND
+
+    def run_all_xspectra_core(self):
+        """Call all ``XspectraCoreWorkChain``s required to compute all requested spectra."""
+
+        structures_to_process = self.ctx.structures_to_process
+        equivalent_sites_data = self.ctx.equivalent_sites_data
+        abs_atom_marker = self.inputs.abs_atom_marker.value
+
+        for site in structures_to_process:
+            inputs = AttributeDict(self.exposed_inputs(XspectraCoreWorkChain, namespace='core'))
+            structure =structures_to_process[site]
+            inputs.structure = structure
+            abs_element = equivalent_sites_data[site]['symbol']
+
+            if 'core_hole_treatments' in self.inputs:
+                ch_treatments = self.inputs.core_hole_treatments.get_dict()
+                ch_treatment = ch_treatments.get(abs_element, 'full')
+            else:
+                ch_treatment = 'full'
+
+            inputs.metadata.call_link_label = f'{site}_xspectra'
+            inputs.eps_vectors = orm.List(list=self.ctx.eps_vectors)
+            if 'core_wfc_data' in self.inputs:
+                inputs.core_wfc_data = self.inputs.core_wfc_data[abs_element]
+            else:
+                inputs.core_wfc_data = self.ctx[f'upf2plotcore_{abs_element}'].outputs.stdout
+
+            # Get the given settings for the SCF inputs and then overwrite them with the
+            # chosen core-hole approximation, then apply the correct pseudopotential pair
+            scf_params = inputs.scf.pw.parameters.get_dict()
+            ch_inputs = XspectraCoreWorkChain.get_treatment_inputs(treatment=ch_treatment)
+            new_scf_params = recursive_merge(left=scf_params, right=ch_inputs)
+
+            # Set the parameter `xiabs` correctly, now that we know which structure we have
+            new_xs_params = inputs.xs_prod.xspectra.parameters.get_dict()
+            kinds_present = sorted([kind.name for kind in structure.kinds])
+            abs_species_index = kinds_present.index(abs_atom_marker) + 1
+            new_xs_params['INPUT_XSPECTRA']['xiabs'] = abs_species_index
+
+            # Set `starting_magnetization` if we are using an XCH approximation, using the
+            # absorbing species as a reasonable place for the unpaired electron.
+            if ch_treatment == 'xch_smear':
+                new_scf_params['SYSTEM'][f'starting_magnetization({abs_species_index})'] = 1
+
+            core_hole_pseudo = self.inputs.core_hole_pseudos[abs_element]
+            gipaw_pseudo = self.inputs.gipaw_pseudos[abs_element]
+            inputs.scf.pw.pseudos[abs_atom_marker] = core_hole_pseudo
+            inputs.scf.pw.pseudos[abs_element] = gipaw_pseudo
+
+            inputs.scf.pw.parameters = orm.Dict(dict=new_scf_params)
+            inputs.xs_prod.xspectra.parameters = orm.Dict(dict=new_xs_params)
```

Comment:
Lose data provenance.

Reply:
The issue here is that AiiDA (as far as I know) doesn't allow the `expose_inputs` port type to be dynamic. Since we need to change the xspectra.x and pw.x input parameters to fit differences in the absorbing atom position, this will technically break data provenance. This is just a consequence of having a `WorkChain` process which branches into multiple sub-processes in order to compute one final property - the same issue exists with the `CoreWorkChain` since we run multiple XAS calculations using one set of "parent" input parameters and thus break provenance for the `parameters`.

If it is possible to have a dynamic set of namespaces for each copy of the input parameters, then we would need to refactor both the `CrystalWorkChain` and `CoreWorkChain` to have a different namespace for each element and each polarisation vector before launching. This could be possible for the `CrystalWorkChain`, since we can know the list of elements to run calculations for beforehand, but not for the `CoreWorkChain` since we won't know the symmetry properties beforehand.